space_packet_parser/xtcedef.py

```diff
@@ -1379,58 +1383,23 @@ def _get_format_string(self, packet_data: PacketData, parsed_data: dict) -> Tupl
-            termination_char_bytes = bytes.fromhex(self.termination_character)
-
-            if self.encoding.startswith("UTF-32"):
-                bytes_per_char = 4
-            elif self.encoding.startswith("UTF-16"):
-                bytes_per_char = 2
-            elif self.encoding in ("UTF-8", "US-ASCII", "ISO-8859-1", "Windows-1252"):
-                bytes_per_char = 1
-            else:
-                raise ValueError(f"Got encoding={self.encoding}. Encoding must be one of {self._supported_encodings}")
-
-            bits_per_byte = 8
-
-            # If we are starting mid-byte, there may not be an integer number of bytes between the string and the
-            # end of the packet
-            n_full_bytes_left_in_packet = (len(packet_data) - packet_data.pos) // bits_per_byte
-
-            # Start by looking ahead at most 16 chars.
-            look_ahead_n_chars = min(16, n_full_bytes_left_in_packet // bytes_per_char)
-
-            # Peek at 64 characters at a time. 64, 128, 192, ...
-            # Remember in utf-16-be/le, each character will be 2 bytes
-            look_ahead_n_bytes = bytes_per_char * look_ahead_n_chars
-
-            while look_ahead_n_bytes <= n_full_bytes_left_in_packet:
-                # Outputs byte string (encoding is irrelevant, they are just bytes)
-                look_ahead = packet_data.peek(f'bytes:{look_ahead_n_bytes}')
-                # Check if the termination character byte string is in the look ahead bytes we peeked at
-                if termination_char_bytes in look_ahead:
-                    # Implicit assumption of one termination character in specified encoding
-                    tclen_bits = bytes_per_char * bits_per_byte
-                    # Split the look ahead bytes string at the termination character and get its length (in bits)
-                    strlen_bits = len(look_ahead.split(termination_char_bytes)[0]) * bits_per_byte
-                    # Tell the parser to skip the termination character
-                    skip_bits_after = tclen_bits
-                    break
-                # Increment look ahead length by one char (1 or 2 bytes, based on encoding)
-                look_ahead_n_bytes += bytes_per_char * look_ahead_n_chars
-                # Ensure we never go over the total length of the packet, in bits
-                look_ahead_n_bytes = min(look_ahead_n_bytes, n_full_bytes_left_in_packet)
-            else:
-                raise ValueError(f"Reached end of binary string without finding "
-                                 f"termination character {self.termination_character}.")
+            # Look through the rest of the packet data to find the termination character
+            nbits_left = len(packet_data) - packet_data.pos
+            string_buffer = packet_data.peek(f"bytes:{nbits_left//8}")
+            try:
+                strlen_bits = string_buffer.index(self.termination_character) * 8
```

Comment:
```suggestion
                strlen_bits = string_buffer.index(bytes.fromhex(self.termination_character)) * 8
```

Reply:
This is one of the places where I put the `bytes.fromhex()` call up into the `__init__` function so it is just called once per packet type rather than each time we read a packet. (`self.termination_character` is the bytes representation at this point)

Reply:
That works for me.